modules/tinymce/src/core/main/ts/selection/MultiClickSelection.ts

```diff
@@ -1,44 +1,82 @@
 import { SugarElement } from '@ephox/sugar';
 
+import DomTreeWalker from '../api/dom/TreeWalker';
 import Editor from '../api/Editor';
-import * as CaretFinder from '../caret/CaretFinder';
+import { isCaretCandidate } from '../caret/CaretCandidate';
 import { CaretPosition } from '../caret/CaretPosition';
-import { isTextBlock } from '../dom/ElementType';
+import { isBlock } from '../dom/ElementType';
 import * as NodeType from '../dom/NodeType';
 import * as RangeNormalizer from './RangeNormalizer';
 
-const isTextBlockNode = (node: Node): node is Element => NodeType.isElement(node) && isTextBlock(SugarElement.fromDom(node));
+const isBr = NodeType.isBr;
+const isText = NodeType.isText;
 
-const normalizeSelection = (editor: Editor) => {
-  const rng = editor.selection.getRng();
-  const startPos = CaretPosition.fromRangeStart(rng);
-  const endPos = CaretPosition.fromRangeEnd(rng);
+const isBoundary = (node: Node): boolean => {
+  return isBr(node) || isBlock(SugarElement.fromDom(node));
+};
 
-  if (CaretPosition.isElementPosition(startPos)) {
-    const container = startPos.container();
-    if (isTextBlockNode(container)) {
-      CaretFinder.firstPositionIn(container).each((pos) => rng.setStart(pos.container(), pos.offset()));
+const walkBackwardWhile = (startNode: Node, rootNode: Node): Node => {
+  const walker = new DomTreeWalker(startNode, rootNode);
+  let result: Node = startNode;
+  for (let next = walker.prev(); next && !isBoundary(next); next = walker.prev()) {
+    if (isCaretCandidate(next)) {
+      result = next;
     }
   }
+  return result;
+};
 
-  if (CaretPosition.isElementPosition(endPos)) {
-    const container = startPos.container();
-    if (isTextBlockNode(container)) {
-      CaretFinder.lastPositionIn(container).each((pos) => rng.setEnd(pos.container(), pos.offset()));
+const walkForwardWhile = (startNode: Node, rootNode: Node): Node => {
+  const walker = new DomTreeWalker(startNode, rootNode);
+  let result: Node = startNode;
+  for (let next = startNode; next && !isBoundary(next); next = walker.next()) {
+    if (isCaretCandidate(next)) {
+      result = next;
     }
   }
+  return result;
+};
+
+const findClosestBlockRange = (startRng: Range, rootNode: Node) => {
+  const startPos = CaretPosition.fromRangeStart(startRng);
+  let clickNode = startPos.container();
```

Comment:
I see that startPos.getNode() function uses deprecated RangeNodes.getNodeUnsafe function. Is the current moment the appropriate time to remove it and replace with RangeNodes.getNode() function?

Reply:
Yeah it does rely on it returning undefined. When I first added the newer/safer `getNode` function I swapped that but then a number of tests failed as it's relying on the unsafe behaviour still. As such given time constraints, I renamed the original function to include unsafe so we hopefully wouldn't use it in future and left this using the older logic.

Reply:
If I recall correctly from looking at it sometime later it was https://github.com/tinymce/tinymce/blob/236891ffcdf2fd1b698d020e67c31838fb110019/modules/tinymce/src/core/main/ts/caret/CaretPosition.ts#L203 that relied on it returning `undefined` as the index ended up being `-1`. As such, the safe method constrained that back to `0` and returned a `node`, whereas the unsafe version returned `undefined`.

I could be remembering incorrectly though, but yeah if we're concerned about it I'd suggest logging a separate Jira and handling the changes there.

Reply:
Actually, I'm not sure `getNodeUnsafe` should be deprecated, as we can see it has its own place. It might have certain usage, which depends on desired logic.